apps/labs/posts/python-wheels-from-tags-to-variants.md

```diff
@@ -0,0 +1,1146 @@
+to be ordered by the plugin, then optionally reordered by the package and the
+user. Finally, all three keys combine into a sort key." />
+  <figcaption>Fig. 8. Ordering and construction of a property sort key</figcaption>
+</figure>
+</div>
+
+Unsurprisingly, plugins provide the initial ordering of features and their
+values. After all, given that the plugin defines these lists in the first place,
+it is quite reasonable for it to order them as well. Otherwise, the package
+maintainers would have to keep repeating that `12.8` > `12.7` > `12.6` > …, and updating this
+list whenever they are about to use a new property value.
+However, plugins are scoped to themselves and can only order feature names
+and values. They cannot provide namespace ordering, as that would effectively
+mean one plugin deciding how important another plugin is.
+
+This leaves us with two possibilities: either the package author or the user
+needs to define namespace ordering. Originally, we went with the latter idea;
+after all, the users know best whether they prefer CUDA or ROCm, or perhaps CPU
+optimization. However, this meant that the user had to jump through
+the hoops of configuring variant usage first, and once again things could not
+work out of the box. So we went with the next best thing possible: requiring package
+maintainers to specify namespace ordering in the variant configuration; after all, package authors
+are in good position to specify which variants are the most beneficial to their software.
+
+<figure>
+
+```toml
+[variant.default-priorities]
+# prefer aarch64 over x86_64
+namespace = ["aarch64", "x86_64"]
+# prefer variants using hardware AES optimizations over these for a newer CPU
+# architecture
+feature.aarch64 = ["aes"]
+feature.x86_64 = ["aes"]
+# prefer x86-64-v3 and then older (even if the CPU is newer)
+property.x86_64.level = ["v3", "v2", "v1"]
+
+[variant.providers.aarch64]
+# Using different package based on the Python version
+requires = ["provider-variant-aarch64 >=0.0.1,<1"]
+# use only on aarch64/arm machines
+enable-if = "platform_machine == 'aarch64' or 'arm' in platform_machine"
+plugin-api = "provider_variant_aarch64.plugin:AArch64Plugin"
+
+[variant.providers.x86_64]
+requires = ["provider-variant-x86-64 >=0.0.1,<1"]
+# use only on x86_64 machines
+enable-if = "platform_machine == 'x86_64' or platform_machine == 'AMD64'"
+plugin-api = "provider_variant_x86_64.plugin:X8664Plugin"`
+```
+
+<figcaption>Listing 3. Example `pyproject.toml` with sort order defined</figcaption>
+</figure>
+
+However, no reason to stop at namespace. After all, a specific package may
+want to point out, say, that their AES-NI variant is more performant than
+a fallback implemented using the instruction sets provided by x86-64-v3.
+It might also be desirable reorder value; while
+I immediately can't think of a specific use case for that, let's have the option
+for symmetry anyway. So while the plugins specify the initial order of features
```

Comment:
"let's have the option anyway" - not clear to me what is being talked about here. Is this in addition to the toml config just shown, or just describing it after the fact?

Reply:
Uh, I feel like I've been switching between past to present tense too much.